juju/applications.go

```diff
@@ -0,0 +1,560 @@
+		aux := charmresources.Resource{
+			Meta: charmresources.Meta{
+				Name:        v.Name,
+				Type:        v.Type,
+				Path:        v.Path,
+				Description: v.Description,
+			},
+			Origin: charmresources.OriginStore,
+			// TODO: prepare for resources with different versions
+			Revision: -1,
+		}
+		pendingResources = append(pendingResources, aux)
+	}
+
+	resourcesReq := apiresources.AddPendingResourcesArgs{
+		ApplicationID: appName,
+		CharmID: apiresources.CharmID{
+			URL:    charmID.URL,
+			Origin: charmID.Origin,
+		},
+		CharmStoreMacaroon: nil,
+		Resources:          pendingResources,
+	}
+
+	toRequest, err := resourcesAPIClient.AddPendingResources(resourcesReq)
+	if err != nil {
+		return nil, err
+	}
+
+	// now build a map with the resource name and the corresponding UUID
+	toReturn := map[string]string{}
+	for i, argsResource := range pendingResources {
+		toReturn[argsResource.Meta.Name] = toRequest[i]
+	}
+
+	return toReturn, nil
+}
+
+func (c applicationsClient) ReadApplication(ctx context.Context, input *ReadApplicationInput) (*ReadApplicationResponse, error) {
+	conn, err := c.GetConnection(ctx, &input.ModelUUID)
+	if err != nil {
+		return nil, err
+	}
+
+	applicationAPIClient := apiapplication.NewClient(conn)
+	defer applicationAPIClient.Close()
+
+	charmsAPIClient := apicharms.NewClient(conn)
+	defer charmsAPIClient.Close()
+
+	clientAPIClient := apiclient.NewClient(conn)
+	defer clientAPIClient.Close()
+
+	apps, err := applicationAPIClient.ApplicationsInfo([]names.ApplicationTag{names.NewApplicationTag(input.ApplicationName)})
+	if err != nil {
+		return nil, err
+	}
+	if len(apps) > 1 {
+		return nil, fmt.Errorf("more than one result for application: %s", input.ApplicationName)
+	}
+	if len(apps) < 1 {
+		return nil, fmt.Errorf("no results for application: %s", input.ApplicationName)
+	}
+	appInfo := apps[0].Result
+
+	var appConstraints constraints.Value = constraints.Value{}
+	// constraints do not apply to subordinate applications.
+	if appInfo.Principal {
+		queryConstraints, err := applicationAPIClient.GetConstraints(input.ApplicationName)
+		if err != nil {
+			return nil, err
+		}
+		if len(queryConstraints) != 1 {
+			return nil, fmt.Errorf("expected one set of application constraints, received %d", len(queryConstraints))
+		}
+		appConstraints = queryConstraints[0]
+	}
+
+	status, err := clientAPIClient.Status(nil)
+	if err != nil {
+		return nil, err
+	}
+	var appStatus params.ApplicationStatus
+	var exists bool
+	if appStatus, exists = status.Applications[input.ApplicationName]; !exists {
+		return nil, fmt.Errorf("no status returned for application: %s", input.ApplicationName)
+	}
+
+	unitCount := len(appStatus.Units)
+
+	// NOTE: we are assuming that this charm comes from CharmHub
+	charmURL, err := charm.ParseURL(appStatus.Charm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse charm: %v", err)
+	}
+
+	returnedConf, err := applicationAPIClient.Get("master", input.ApplicationName)
```

Comment:
Where does this use of `"master"` come from? Is this some Juju internal thing?

Reply:
The [method](https://github.com/juju/juju/blob/cbffc09649a05a036caf8cf1253af6f405451e0e/api/client/application/client.go#L709) names this argument branch name. I think juju might use git internally to manage config maybe?